lib/operations/execute_operation.js

```diff
@@ -30,64 +31,57 @@ function executeOperation(topology, operation, callback) {
     throw new TypeError('This method requires a valid operation instance');
   }
 
-  if (isUnifiedTopology(topology) && topology.shouldCheckForSessionSupport()) {
-    return selectServerForSessionSupport(topology, operation, callback);
-  }
-
-  const Promise = topology.s.promiseLibrary;
-
-  // The driver sessions spec mandates that we implicitly create sessions for operations
-  // that are not explicitly provided with a session.
-  let session, owner;
-  if (topology.hasSessionSupport()) {
-    if (operation.session == null) {
-      owner = Symbol();
-      session = topology.startSession({ owner });
-      operation.session = session;
-    } else if (operation.session.hasEnded) {
-      throw new MongoError('Use of expired sessions is not permitted');
+  return maybePromise(topology, cb, callback => {
+    if (isUnifiedTopology(topology) && topology.shouldCheckForSessionSupport()) {
```

Comment:
sorry to be so nitpicky, but I prefer what we do in the [4.0 branch](https://github.com/mongodb/node-mongodb-native/blob/4.0/src/operations/execute_operation.ts#L64-L75) where there's a `maybePromise` for each of these cases. The way this is written right now, you might end up nesting promises, whereas in 4.0 if you haven't provided a callback we are using the callback from the top `maybePromise` to pass into the secondary `executeOperation`.

Reply:
(No worries, please nitpick away!)
I don't see how you can nest promises here since `selectServerForSessionSupport` will get passed a defined callback from `maybePromise` and so the next call to executeOperation will have that callback defined so the "second" `maybePromise` won't create a new Promise.

Something to consider is that this provides one point of exit to executeOperation. Entirely an 🧅  but that pattern is preferable to what is on 4.0. It's a strange pattern to me to have more code run "synchronously" in the remainder of the function after you've returned a function that appears to do an async task. (In this instance, I know maybePromise is all sync it's just wrapping a bit of setup work, but my style point remains). I'm glossing over the callback detail by just returning but we get the idea. 
```js
function runOp() {
  if (cond)
    return doThing(() => {
      // now we've done something
      return 'one way'
    })

  // do more things
  return doThing(() => {
      // now we've done something
      return 'the other way'
  })
}
```
vs.
```js
function runOp() {
  return doThing(() => {
    if (cond) return 'one way'
    
    // do more things
    return 'the other way'
  })
}
```

Reply:
Hm, you're right there's no Promise nesting. Maybe my ultimate concern is just about the multiple `maybePromise` calls in the `executeOperation` stacktrace. A call to `insertOne` right now will look like:

Reply:
Yea we'll need to make a port of this sessions related fix to 4.0 so I can do that when I'm there (If anyone else reviewing has a differing opinion please feel free to chime in!)